File: hllpp_py/src/hllpp/compare3.py
############################################
# Copyright (c) 2026 Shun/修海 (@shun4midx) #
# Project: HyperLogLogPlusPlus-Autocorrect #
# File Type: Python file                   #
# File: compare3.py                        #
############################################

import sys

def compare3_files(file1, file2, ground_truth):
    with open(file1, 'r', encoding='utf-8') as f1, \
         open(file2, 'r', encoding='utf-8') as f2, \
         open(ground_truth, 'r', encoding='utf-8') as fg:
        out1 = [line.strip() for line in f1]
        out2 = [line.strip() for line in f2]
        gold = [line.strip() for line in fg]

    max_len = max(len(out1), len(out2), len(gold))
    out1 += [''] * (max_len - len(out1))
    out2 += [''] * (max_len - len(out2))
    gold += [''] * (max_len - len(gold))

    correct1 = correct2 = 0

    print(f"{'Idx':<5} | {'Truth':<20} | {'File 1 (top3)':<30} | {'File 2':<20} | Result")
    print("-" * 100)

    for i in range(max_len):
        t = gold[i]
        a = out1[i]
        b = out2[i]

        # File1 may have up to 3 suggestions separated by spaces
        suggestions1 = a.split()
        suggestions2 = b.split()
        result = []
        if t in suggestions1:
            correct1 += 1
            result.append("OK1")
        if t in suggestions2:
            correct2 += 1
            result.append("OK2")

        print(f"{i+1:<5} | {t:<20} | {a or '[empty]':<50} | {b or '[empty]':<20} | {' '.join(result) or 'X'}")

    print("\nSummary")
    print("-" * 100)
    print(f"Total queries     : {max_len}")
    print(f"Correct (File 1)  : {correct1} ({correct1/max_len*100:.2f}%)")
    print(f"Correct (File 2)  : {correct2} ({correct2/max_len*100:.2f}%)")

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python compare3.py suggestions.txt typos.txt true_answers.txt")
        sys.exit(1)

    compare3_files(sys.argv[1], sys.argv[2], sys.argv[3])

File: hllpp_cpp/src/src/compare3.cpp
/********************************************
 * Copyright (c) 2026 Shun/修海 (@shun4midx) *
 * Project: HyperLogLogPlusPlus-Autocorrect *
 * File Type: C++ file                      *
 * File: compare3.cpp                       *
 ****************************************** */

// ======== INCLUDE ======== //
#include "../include/HLLPP/compare3.h"

// ======== FUNCTION IMPLEMENTATION ======== //
static std::unordered_set<std::string> split(const std::string& s) {
    std::istringstream iss(s);
    std::unordered_set<std::string> tokens;
    std::string token;

    while (iss >> token) {
        tokens.insert(token);
    }

    return tokens;
}

void compare3_files(std::filesystem::path file1, std::filesystem::path file2, std::filesystem::path ground_truth) {
    // Read all three files into string vectors
    auto read_lines = [](const std::filesystem::path& p){
        std::ifstream in(p);
        if (!in) {
            throw std::runtime_error("Unable to open " + p.string());
        }

        std::vector<std::string> lines;
        std::string line;

        auto trim = [](std::string s) {
            const auto first = s.find_first_not_of(" \t\r\n");
            if (first == std::string::npos) {
                return std::string{};
            }
        
            const auto last = s.find_last_not_of(" \t\r\n");
            return s.substr(first, last - first + 1);
        };

        while (std::getline(in, line)) {
            lines.push_back(trim(line));
        }

        return lines;
    };

    std::vector<std::string> out1 = read_lines(file1);
    std::vector<std::string> out2 = read_lines(file2);
    std::vector<std::string> gold = read_lines(ground_truth);

    // Pad all three to the same length
    int max_len = std::max(std::max(out1.size(), out2.size()), gold.size());
    out1.resize(max_len, "");
    out2.resize(max_len, "");
    gold.resize(max_len, "");

    // Compare
    int correct1 = 0, correct2 = 0;

    // Header
    std::cout << std::left << std::setw(5)  << "Idx" << " | " << std::setw(20) << "Truth" << " | " << std::setw(30) << "File 1 (top3)" << " | " << std::setw(20) << "File 2" << " | Result\n"
              << std::string(100, '-') << "\n";

    for (int i = 0; i < max_len; ++i) {
        auto& t = gold[i];
        auto& a = out1[i];
        auto& b = out2[i];

        std::unordered_set<std::string> suggestions1 = split(a);
        std::unordered_set<std::string> suggestions2 = split(b);
        std::vector<std::string> result;

        if (suggestions1.find(t) != suggestions1.end()) {
            ++correct1;
            result.push_back("OK1");
        }

        if (suggestions2.find(t) != suggestions2.end()) {
            ++correct2;
            result.push_back("OK2");
        }

        auto disp = [&](const std::string& s){
            return s.empty() ? "[empty]" : s;
        };

        std::string joined_result = "";

        for (std::string& str : result) {
            if (!joined_result.empty()) {
                joined_result += " ";
            }
            joined_result += str;
        }

        std::cout << std::setw(5) << (i + 1) << " | " << std::setw(20) << disp(t) << " | " << std::setw(50) << (!a.empty() ? a : "[empty]") << " | " << std::setw(20) << (!b.empty() ? b : "[empty]") << " | " << (!joined_result.empty() ? joined_result : "X") << "\n";
    }

    // Summary
    std::cout << "\nSummary\n" << std::string(100, '-') << "\n"
              << std::fixed << std::setprecision(2)
              << "Total queries            : " << max_len << "\n"
              << "Correct in File 1        : " << correct1 << " (" << (double)correct1 / max_len * 100 << "%)\n"
              << "Correct in File 2        : " << correct2 << " (" << (double)correct2 / max_len * 100 << "%)\n";
}